### PR TITLE
ISSUE-680 # Add the DB SQL Executor to import data from CSV and execute SQL statements

### LINKED ISSUE #680: [DB] Import or load data from an external CSV file to a table
# AC1:
Given a datafile in CSV format, matching the Database table columns,
provide mechanism to load these data(rows and columns) in the that table.

Example:
- Table "players" has three columns as below:
```
ID:
NAME:
AGE:
```

and data in CSV is as below:
```
1001, Ronaldo, 23
1002, Devaldo, 24
1003, Trevaldo, 35
```
Note- There are no headers present in the above CSV

# AC2:
Same as AC1, but with headers.

Example:

and data in CSV is as below:

```
ID, AGE, NAME <----------- 1st row as header
1001, 23, Ronaldo
1002, 24, Devaldo
1003, 35, Trevaldo
```

# AC3 :
Required for Postgres DB (v9 to v12 should be supported )

# AC4: 
File could be present in the local files system(eg /test/resources) 

optional:
(or in a git repo ==> implement this via a separate ticket.)

# AC5 : 
```JSON
{
    "scenarioName": "Sample interaction with the db executor ( data from AC2)",
    "steps": [
        {
            "name": "Loading a csv file",
            "url": "org.jsmart.zerocode.core.db.DbSqlExecutor",
            "operation": "LOADCSV", //<--- in uppercase (for consistency), but support both cases
            "request": {
                "tableName": "players",
                "csvSource": "path/db_sample.csv",
                "withHeaders" : false, //<--false:  this is the default value most cases. (optional)
                "nullString" : "" //<-- used to insert null values, this is the default value (optional)
            },
            "verify": { 
                "size" : 36 //<-- 36 records: number of records loaded. (optional field as usual)
             } 
       },
        {
            "name": "assert_loaded_data_via_select",
            "url": "org.jsmart.zerocode.core.db.DbSqlExecutor",
            "operation": "EXECUTE", //<---- executes an in-line SQL (in this case a SELECT sql)
            "request": {
                "sql": "SELECT id, name, age FROM players WHERE age > ? ORDER BY age ASC",
                "sqlParams" : [ 23 ]
            },
            "verify": {
                "rows.SIZE": 2, //<--- Number of rows in the output (This is OOTB available, don't have to implement)
                "rows": [
                    {
                        "id": 1002,
                        "name": "Devaldo",
                        "age": 24
                    },
                    {
                        "id": 1003,
                        "name": "Trevaldo",
                        "age": 35
                    }
                ]
            }
        }
    ]
}
```

# AC6: 

_From @javiertuya :_
This is a proposal for discussion before starting the implementation:

- Where to place the code? Ans: `org.jsmart.zerocode.core.db` package

- Use Apache Commons DbUtils that is a very stable and lightweight library. 
  This will allow to implement something that does not depend on any particular DBMS.
- Add an executor (e.g. `DbSqlExecutor`) with the operation  explained in the example
  as well as `EXECUTE` (or `SELECT` and `UPDATE` separately) to handle SQL queries with parameters
- Tests would use H2, as it is already included as test scoped dependency. 
- Postgress requires a container(an example of Docker already present), this should be tested manually, or added to the CI pipeline if required.

More >>
- CSV reading will be based/inspired on the domain `Parametrized.java` class. : What does this mean? Please explain this or point to the PR when implemented.

On the **AC6** or **AC5**, please discuss with @javiertuya if you have any questions or suggestions.

ACRONYMS:
-------------
- OOTB : Out Of The Box

### PR DESCRIPTION
**Pull request by javiertuya**
# Add the DB SQL Executor to import data from CSV and execute SQL statements

## Fixes Issue
- [x] Which issue or ticket was(will be) fixed by this PR? Close #680 

PR Branch
**https://github.com/javiertuya/zerocode/tree/680-db-sql-executor**

Documentation at PR https://github.com/authorjapps/zerocode-tdd-docs/pull/24

## Motivation and Context

This PR provides an out of the box component that uses the Java API to execute steps that interact with a database.

The executor class `org.jsmart.zerocode.core.db.DbSqlExecutor` provides two operations:
- `EXECUTE`: Run a SQL statement to verify the data stored in the database or insert/update in the data.
- `LOADCSV`: Load the contents of a CSV file located in the resources folder into a table.

Acceptance critera fulfillment:
- AC1, AC2, AC5, AC6: Done, with unit and integration tests, H2 database.
- AC3: Done, PostgreSQL: the same unit and integration tests were executed in local, not in CI
- AC4: Partially done, read CSV files from src/test/resources. Reading files from URL to be implemented in separate issue

Suggestions for future improvements:
- Load from an url (AC4)
- Load from sites that require authentication (mentioned in the issue)
- Load from anywhere in the filesystem
- Custom datetime formats (specified in the properties file or in the request)
- Load data into binary columns (e.g. hexadecimal, base64...)
- Run the PostgreSQL tests as part of the GitHub Actions workflow

## Checklist:

* [x] New Unit tests were added
  * [ ] Covered in existing Unit tests

* [x] Integration tests were added
  * [ ] Covered in existing Integration tests

* [x] Test names are meaningful

* [x] Feature manually tested and outcome is successful

* [x] PR doesn't break any of the earlier features for end users
  * [ ] WARNING! This might break one or more earlier earlier features, hence left a comment tagging all reviewrs

* [x] Branch build passed in CI

* [x] No 'package.*' in the imports

* [x] Relevant DOcumentation page added or updated with clear instructions and examples for the end user
  * [ ] Not applicable. This was only a code refactor change, no functional or behaviourial changes were introduced

* [ ] Http test added to `http-testing` module(if applicable) ?
  * [x] Not applicable. The changes did not affect HTTP automation flow

* [ ] Kafka test added to `kafka-testing` module(if applicable) ?
  * [x] Not applicable. The changes did not affect Kafka automation flow

### CODE REVIEW

**Comment by nirmalchandra** [comment]
@javiertuya , excellent work!
I've given the CI run and awaiting for a Green(check status [here](https://github.com/authorjapps/zerocode/pull/686/checks)).

Just checking:
1. Did you get chance to run a Integration-test scenario against a real Postgres DB (as the CI tests are using h2db) ?
   - Even Docker should be fine I guess, as this will validate the JDBC driver aspects ok 
2. Is it possible for you(if you by chance have an Infra) to run against a remote DB instance (example: RDS ) to check the Drivers work as expected?

**Comment by javiertuya** [comment]
@nirmalchandra 
> 1. Did you get chance to run a Integration-test scenario against a real Postgres DB (as the CI tests are using h2db) ?

Yes. The target DB is configured in `core/src/test/resources/db_test.properties` (this file has some instructions embedded). I ran both **unit and integration** tests using three different versions of Postgres. The only difference is that Postgress stores column and table identifiers as lowercase, so that, to make the integration tests pass, you need convert the json scenarios to have the key values in the `verify` to lowercase (this is done from the junit in the case of unit tests). For example, to run the integration test that loads a table from a CSV file with headers, you can do this case conversion by running from the `db` subfolder:
```
sed -i 's/"ID"/"id"/' db_csv_load_with_headers.json
sed -i 's/"NAME"/"name"/' db_csv_load_with_headers.json
sed -i 's/"AGE"/"age"/' db_csv_load_with_headers.json
```

>    
>    * Even Docker should be fine I guess, as this will validate the JDBC driver aspects ok

It would also be possible to run the postgres tests in the Github Actions (spin-up postgres takes around 6 seconds)

> 2. Is it possible for you(if you by chance have an Infra with any DB) to run against a remote DB instance (example: RDS ) to check the Drivers work as expected?

I just checked it: connecting from a desktop to a remote server in my lab (through a VPN) and it works fine. To check this, just run a postgress instance in the server, and form the desktop, change the `db.driver.url` property to put the server ip and port

**Comment by authorjapps** [comment]
@a1shadows , @omkar-shitole 👋 : 
Can you guys also have a look at the PR and review it?
Also if possible, check against some other real DB e.g. real MySQL etc?

**Comment by a1shadows** [comment]
Should we also try and create some documentation to https://github.com/authorjapps/zerocode-tdd-docs/tree/main?
Otherwise, feature looks good to me. I'll try and run it on my local this weekend

**Comment by authorjapps** [comment]
> Should we also try and create some documentation to https://github.com/authorjapps/zerocode-tdd-docs/tree/main? Otherwise, feature looks good to me. I'll try and run it on my local this weekend

@a1shadows ,
It's already [here](https://github.com/authorjapps/zerocode/issues/680#issuecomment-2423965269).

But, please check if you want to add anything to it or amend it to make it more helpful, then please proceed 👍

**Comment by authorjapps** [comment]
@javiertuya , I ran this against a Postgres DB. It didn't execute as expected. Can you have a look and fix it if possible?

Looks like it is expecting a correct Driver.

_Note: I didn't create any tables yet, just wanted to see if the SQL gets executed or not._

(Also I tried with latest Postgres version, same error. I think PG version is not an issue here. It's not even trying to validate user/password supplied via config)
Tried this from your config file:
```
#     docker run --name zerocode-postgres -p:5432:5432 -e POSTGRES_PASSWORD=mypassword -d postgres:17.0
#     docker run --name zerocode-postgres -p:5432:5432 -e POSTGRES_PASSWORD=mypassword -d postgres:12.0
```

Docker PG brought up from here:
Branch:
> pr_686_db_sql/zerocode/docker/compose/pg_compose.yml

 I ran this
```
src/test/resources/integration_test_files/db/db_sql_execute.json
```

Config:
test/resources/db_test.properties :
```
db.driver.url=jdbc:postgresql://localhost:35432/postgres
db.driver.user=postgres
db.driver.password=example
```


Pasted Graphic

**Comment by javiertuya** [comment]
@authorjapps Done. A new commit that:
- Adds the new postgres test that runs two scenarios in CI (load csv with headers and execute sql).
- Adds the properties file (pg driver configuration) and the two json scenarios (as postgres stores identifiers in lowercase, they are a copy of the corresponding H2 tests, with the assert keys in lowercase).
- Naming convention: files related to the pg tests add the word postgres as a suffix.
- Comments updated in other files referring to running the postgres tests.

**Comment by authorjapps** [comment]
We may not need this in `test` scope.
We need this Driver version `42.7.4` be available as default in the dependant projects.
_(Whichever project needs a different version, they can override in their POM file.)_

> Driver <postgres.db.version>42.7.4</postgres.db.version> 
   ==> runs fine against `postgres:9.3` Db
(Let me create a table to list this Postgres DB vs Driver compatible, which might be useful later on.

```
<dependency> 
                <groupId>org.postgresql</groupId>
                <artifactId>postgresql</artifactId>
                <version>${postgres.db.version}</version>
<!--                <scope>test</scope>-->
</dependency>
```

**Review by nirmalchandra — APPROVED** [review]

**Comment by javiertuya** [comment]
@authorjapps @nirmalchandra This PR is approved and merged 👋, but there is an unresolved comment: https://github.com/authorjapps/zerocode/pull/686#discussion_r1792532640.
Do you still want me to move the integration tests as indicated in this comment?

**Comment by authorjapps** [comment]
@javiertuya , yes please, appreciate you spotted this. Thanks.
A new PR would should do.

Cheers

File: core/src/test/java/org/jsmart/zerocode/core/db/DbSqlExecutorScenarioTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package org.jsmart.zerocode.core.db;
+import org.jsmart.zerocode.core.domain.Scenario;
+import org.jsmart.zerocode.core.domain.TargetEnv;
+import org.jsmart.zerocode.core.runner.ZeroCodeUnitRunner;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+@TargetEnv("db_test.properties")
+@RunWith(ZeroCodeUnitRunner.class)
+public class DbSqlExecutorScenarioTest {
```
[the remark this answers is not in the document]
Comment (javiertuya) [reply core/src/test/java/org/jsmart/zerocode/core/db/DbSqlExecutorScenarioTest.java:10]: Addressed in #698

**Pull request merged**